### DISCUSSION TITLE
Canvas path hole (counter shape)

### DISCUSSION THREAD

**Discussion by cdaein** [thi-ng/umbrella#419] [Q&A] [answered]
I was trying to create a shape with a hole. With Canvas API, I draw the main shape clockwise, and the counter shape counter-cw.
But it seems `PathBuilder` from `@thi.ng/geom` works differently.

If anyone has a tip, I'd appreciate it.

```js
// prettier-ignore
const pts: [number, number][] = [
  [100, 100], [500, 100], [500, 500], [100, 500]
];

// prettier-ignore
const hole: [number, number][] = [
  [300, 200], [200, 400], [400, 400] 
];

// canvas api
pts.forEach((pt, i) => (i === 0 ? ctx.moveTo(...pt) : ctx.lineTo(...pt)));
hole.forEach((pt, i) => (i === 0 ? ctx.moveTo(...pt) : ctx.lineTo(...pt)));
ctx.fillStyle = "yellow";
ctx.fill();

// thi.ng
const path = new PathBuilder({ fill: "white", translate: [200, 200] });
pts.forEach((pt, i) => (i === 0 ? path.moveTo(pt) : path.lineTo(pt)));
hole.forEach((pt, i) => (i === 0 ? path.moveTo(pt) : path.lineTo(pt)));
draw(ctx, path.current());
```

![2023 10 11-14 55 20](https://github.com/thi-ng/umbrella/assets/3522003/1a33a51d-d1e8-4814-8b74-20debc563785)

**Comment by postspectacular** [comment, marked as answer]
Hey @cdaein - sorry, just saw your question now! Alas, holes are currently not directly supported, neither in paths nor in polygons. For various reasons, only simple boundary representations for now...

**Reply by cdaein** [reply]
Got it. Thank you for the response!
